relay.go

```diff
@@ -17,8 +24,96 @@ type RelayHosts interface {
 }
 
 type relayItem struct {
+	*time.Timer
+
 	remapID     uint32
 	destination *Relayer
+	tomb        bool
+}
+
+type relayItems struct {
+	sync.RWMutex
+
+	logger Logger
+	tombs  uint64
+	items  map[uint32]relayItem
+}
+
+func newRelayItems(logger Logger) *relayItems {
+	return &relayItems{
+		items: make(map[uint32]relayItem),
+	}
+}
+
+// CountAll returns the number of items in the relay, including tombstones.
+func (r *relayItems) CountAll() int {
+	r.RLock()
+	n := len(r.items)
+	r.RUnlock()
+	return n
+}
+
+// Count returns the number of non-tombstone items in the relay.
+func (r *relayItems) Count() int {
+	r.RLock()
+	n := len(r.items)
+	tombs := int(r.tombs)
+	r.RUnlock()
+	return n - tombs
+}
+
+// Get checks for a relay item by ID, returning the item and a bool indicating
+// whether the item was found.
+func (r *relayItems) Get(id uint32) (relayItem, bool) {
+	r.RLock()
+	item, ok := r.items[id]
+	r.RUnlock()
+
+	return item, ok
+}
+
+// Add adds a relay item.
+func (r *relayItems) Add(id uint32, item relayItem) {
+	r.Lock()
+	r.items[id] = item
+	r.Unlock()
+}
+
+// Delete removes a relayItem completely (without leaving a tombstone).
+func (r *relayItems) Delete(id uint32) {
+	r.Lock()
+	if item, ok := r.items[id]; ok {
+		item.Stop()
```

Comment:
what do you think about doing all extra operations on the tombstone outside of the lock?